server/schedule/checker/rule_checker.go

```diff
@@ -0,0 +1,277 @@
+	if newScore <= oldScore {
+		log.Debug("no better peer", zap.Uint64("region-id", region.GetID()), zap.Float64("new-score", newScore), zap.Float64("old-score", oldScore))
+		return nil, nil
+	}
+	checkerCounter.WithLabelValues("rule_checker", "move-to-better-location").Inc()
+	newPeer := &metapb.Peer{StoreId: newPeerStore.GetID(), IsLearner: oldPeer.IsLearner}
+	return operator.CreateMovePeerOperator("move-to-better-location", c.cluster, region, operator.OpReplica, oldPeer.GetStoreId(), newPeer)
+}
+
+func (c *RuleChecker) fixOrphanPeers(region *core.RegionInfo, fit *placement.RegionFit) (*operator.Operator, error) {
+	if len(fit.OrphanPeers) == 0 {
+		return nil, nil
+	}
+	checkerCounter.WithLabelValues("rule_checker", "remove-orphan-peer").Inc()
+	peer := fit.OrphanPeers[0]
+	return operator.CreateRemovePeerOperator("remove-orphan-peer", c.cluster, 0, region, peer.StoreId)
+}
+
+func (c *RuleChecker) isDownPeer(region *core.RegionInfo, peer *metapb.Peer) bool {
+	for _, stats := range region.GetDownPeers() {
+		if stats.GetPeer().GetId() != peer.GetId() {
+			continue
+		}
+		storeID := peer.GetStoreId()
+		store := c.cluster.GetStore(storeID)
+		if store == nil {
+			log.Warn("lost the store, maybe you are recovering the PD cluster", zap.Uint64("store-id", storeID))
+			return false
+		}
+		if store.DownTime() < c.cluster.GetMaxStoreDownTime() {
+			continue
+		}
+		if stats.GetDownSeconds() < uint64(c.cluster.GetMaxStoreDownTime().Seconds()) {
+			continue
+		}
+		return true
+	}
+	return false
+}
+
+func (c *RuleChecker) isOfflinePeer(region *core.RegionInfo, peer *metapb.Peer) bool {
+	store := c.cluster.GetStore(peer.GetStoreId())
+	if store == nil {
+		log.Warn("lost the store, maybe you are recovering the PD cluster", zap.Uint64("store-id", peer.StoreId))
+		return false
+	}
+	return !store.IsUp()
+}
+
+// SelectStoreToAddPeerByRule selects a store to add peer in order to fit the placement rule.
+func SelectStoreToAddPeerByRule(scope string, cluster opt.Cluster, region *core.RegionInfo, rf *placement.RuleFit, filters ...filter.Filter) *core.StoreInfo {
+	fs := []filter.Filter{
+		filter.StoreStateFilter{ActionScope: scope, MoveRegion: true},
+		filter.NewStorageThresholdFilter(scope),
+		filter.NewLabelConstaintFilter(scope, rf.Rule.LabelConstraints),
+		filter.NewExcludedFilter(scope, nil, region.GetStoreIds()),
+	}
+	fs = append(fs, filters...)
+	store := selector.NewReplicaSelector(getRuleFitStores(cluster, rf), rf.Rule.LocationLabels).
+		SelectTarget(cluster, cluster.GetStores(), fs...)
+	return store
+}
+
+// SelectStoreToReplacePeerByRule selects a store to replace a region peer in order to fit the placement rule.
+func SelectStoreToReplacePeerByRule(scope string, cluster opt.Cluster, region *core.RegionInfo, fit *placement.RegionFit, rf *placement.RuleFit, peer *metapb.Peer, filters ...filter.Filter) *core.StoreInfo {
+	rf2 := removePeerFromRuleFit(rf, peer)
```

Comment:
What purpose for this,  don't want to remove `losematchPeer`?

Reply:
Similar to [replicaChecker.SelectBestReplacementStore](https://github.com/pingcap/pd/blob/9ae11e20321092edd436431b19124bcd4a9968a9/server/schedule/checker/replica_checker.go#L121-L125), `replace` is done by 
*copy region with the old peer removed + add a new peer to the copied region*

The difference is in `ruleChecker`, replace happens inside a rule (remove a peer, add another peer fits the same rule), so the replace process besomes
*copy RuleFit with the old peer removed + add a new peer to the copied RuleFit*

Reply:
got